feat: show violations and mutations as warning

vishal-chdhry opened the pull request:
Signed-off-by: Vishal Choudhary <vishal.choudhary@nirmata.com>
## Explanation
This PR adds the following
- A new field in policy spec called `auditWarn`. It controls if an audit or mutate policy returns a warning
- Audit policies will return a warning on violations when this field is set to true
- Mutate field will return a warning on mutation when this field is set to true.

<!--
In a couple sentences, explain why this PR is needed and what it addresses. This should be an explanation a non-developer user can understand and covers the "why" question. It should also clearly indicate whether this PR represents an addition, a change, or a fix of existing behavior. This explanation will be used to assist in the release note drafting process.

THIS IS MANDATORY.
-->

## Related issue

Closes: #1127
<!--
Please link the GitHub issue this pull request resolves in the format of `Closes #1234`. If you discussed this change
with a maintainer, please mention her/him using the `@` syntax (e.g. `@JimBugwadia`).

If this change neither resolves an existing issue nor has sign-off from one of the maintainers, there is a
chance substantial changes will be requested or that the changes will be rejected.

You can discuss changes with maintainers in the [Kyverno Slack Channel](https://kubernetes.slack.com/).
-->

## Milestone of this PR
`/milestone 1.13.0`
<!--

Add the milestone label by commenting `/milestone 1.2.3`.

-->

## Documentation (required for features)

My PR contains new or altered behavior to Kyverno. 
- [x] I have sent the draft PR to add or update [the documentation](https://github.com/kyverno/website) and the link is:
  <!-- Uncomment to link to the PR -->
  <!-- https://github.com/kyverno/website/pull/123 -->

## What type of PR is this

<!--

> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading white spaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
-->
/kind feature
## Proposed Changes

<!--
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. 

***NOTE***: If this PR results in new or altered behavior which is user facing, you **MUST** read and follow the steps outlined in the [PR documentation guide](pr_documentation.md) and add Proof Manifests as defined below.
-->

### Proof Manifests

Policy: 
```yaml
apiVersion: kyverno.io/v1
kind: ClusterPolicy
metadata:
  name: check-label-app
spec:
  validationFailureAction: Audit
  auditWarn: true
  rules:
  - name: check-label-app
    match:
      resources:
        kinds:
        - Pod
    validate:
      message: The label 'app' is required.
      pattern:
        metadata:
          labels:
            app: "?*"
---
apiVersion: kyverno.io/v1
kind: ClusterPolicy
metadata:
  name: add-labels
spec:
  auditWarn: true
  rules:
  - match:
      any:
      - resources:
          kinds:
          - Pod
    mutate:
      patchStrategicMerge:
        metadata:
          labels:
            foo: bar
    name: add-labels
  validationFailureAction: Audit

```
Pod: 
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: test-pod
spec:
  containers:
  - name: nginx
    image: nginx:latest

```
Output
```bash
$ kubectl apply -f pod.yaml --dry-run=server
Warning: policy add-labels.add-labels: mutated Pod/test-pod in namespace default
Warning: policy check-label-app.check-label-app: validation error: The label 'app' is required. rule check-label-app failed at path /metadata/labels/app/
pod/test-pod created (server dry run)
```
<!--
Read and follow the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) for more details first. This section is for pasting your YAML manifests (Kubernetes resources and Kyverno policies) and Kyverno CLI test manifests which allow maintainers to prove the intended functionality is achieved by your PR. Please use proper fenced code block formatting, for example:

# Kubernetes resource

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: roles-dictionary
  namespace: default
data:
  allowed-roles: "[\"cluster-admin\", \"cluster-operator\", \"tenant-admin\"]"
```

# Kyverno CLI test manifest (please see docs for latest manifest format at https://kyverno.io/docs/kyverno-cli/). See kyverno/policies for complete examples of all related test files.

```yaml
name: prepend-image-registry
policies:
  - prepend_image_registry.yaml
resources:
  - resource.yaml
variables: values.yaml
results:
  - policy: prepend-registry
    rule: prepend-registry-containers
    resource: mypod
    # if mutate rule
    patchedResource: patchedResource01.yaml
    kind: Pod
    result: pass
```
-->

## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
-->

- [x] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [x] I have read the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) and followed the process including adding proof manifests to this PR.
- [ ] This is a bug fix and I have added unit tests that prove my fix is effective.
- [ ] This is a feature and I have added CLI tests that are applicable.
- [ ] My PR needs to be cherry picked to a specific release branch which is <replace>.
- [ ] My PR contains new or altered behavior to Kyverno and
  - [ ] CLI support should be added and my PR doesn't contain that functionality.

## Further Comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution
you did and what alternatives you considered, etc...
-->

chipzoller commented:
Awesome! Would really like to see Proof Manifests here of these changes in action.

vishal-chdhry commented:
@chipzoller added

chipzoller commented:
Could you also show a mutation warning, please?

vishal-chdhry commented:
@chipzoller, added! 

Thanks for pointing that out, even though  mutation handler returns the warning in unit testing, it does not get shown in kubectl output, I guess the reason for that might be because the output of mutation phase is not the final result and it gets overwritten in the validation step. I have fixed that in the latest commit.

chipzoller commented:
Good. Some other questions:

1. Have you made sure this is also applicable to `Policy` and not just `ClusterPolicy`?
2. Have you tested multiple rules where both rules would return separate messages? What about if only one out of multiple returns a message?
3. Think about events and logs. What, if anything, should happen here?

vishal-chdhry commented:
> Have you made sure this is also applicable to Policy and not just ClusterPolicy?

Yes

> Have you tested multiple rules where both rules would return separate messages? What about if only one out of multiple returns a message?

If only one returns a message then only that message will be printed

> Think about events and logs. What, if anything, should happen here?

looking into it, I will see if there is a case that might be missing

JimBugwadia commented on api/kyverno/v1/spec_types.go:
```diff
@@ -76,6 +76,12 @@ type Spec struct {
+	// Ignored for enforce policies. The default value is "false".
+	// +optional
+	// +kubebuilder:default=false
+	AuditWarn *bool `json:"auditWarn,omitempty" yaml:"auditWarn,omitempty"`
```
Suggested text:

> AuditWarning enables API response warnings for policy rules with validationFailureAction set to Audit.  Enabling this option will extend admission request processing times. The default value is "false".

realshuting commented on pkg/webhooks/resource/validation/validation.go:
```diff
@@ -120,20 +120,19 @@ func (v *validationHandler) HandleValidation(
 		return false, webhookutils.GetBlockedMessages(engineResponses), nil
 	}
 
-	go v.handleAudit(ctx, policyContext.NewResource(), request, policyContext.NamespaceLabels(), engineResponses...)
```
We still want to process audit policies asynchronously, ideally we need to add audit warn policies to the policy cache and retrieve and apply them during admission review:

https://github.com/kyverno/kyverno/blob/fc316aca858d3f7cf5ef715254100bbce2f80be4/pkg/webhooks/resource/handlers.go#L197

vishal-chdhry replied:
I did not update the policy cache to keep the code more simpler so that there is still a clear boundary between audit and enforce

JimBugwadia replied:
The policy cache should be used for all policies to be executed in the admission controller.....not just for enforce mode policies. 

We can add methods like `Policy.AdmissionReview()` which takes which returns true for policies that should be executed during admission review, or `PolicyCache.Get(policyType, admissionReview)` which returns the appropriate policy set from cache.

chipzoller commented:
Would also recommend a couple more basic Chainsaw tests that check stdout for the warning message, one for mutate and one for validate.

vishal-chdhry commented:
superseded by #10214
